File: src/cli/commands/pic2html.ts
/**
 * pic2html — Convert images to HTML table art
 *
 * Modern TypeScript recreation of foxglove's 2007 Pic2Html.exe (VB.NET).
 * Takes an input image and obfuscates it into an HTML table where each
 * pixel (or block of pixels) becomes a colored <td> with bgcolor.
 *
 * Originally built at home on the D: drive, tested with foxglove@philips.com
 * rendered as a 16-level grayscale HTML table — anti-scraping before anyone
 * called it that. June 3, 2007.
 *
 * @example
 *   npx gsd-skill-creator pic2html image.png
 *   npx gsd-skill-creator pic2html image.png --size 24
 *   npx gsd-skill-creator pic2html image.png --grayscale --levels 16
 *   npx gsd-skill-creator pic2html image.png --max-width 80 --quantize 8
 */

import * as p from '@clack/prompts';
import pc from 'picocolors';
import { readFile, writeFile } from 'fs/promises';
import { existsSync } from 'fs';
import { basename, extname } from 'path';

// ── Types ──

interface Pic2HtmlOptions {
  size: number;
  output?: string;
  grayscale: boolean;
  levels: number;
  maxWidth?: number;
  quantize?: number;
  bg: string;
}

interface RGB {
  r: number;
  g: number;
  b: number;
}

// ── Color math ──

function quantizeColor(r: number, g: number, b: number, levels: number): RGB {
  const step = Math.floor(256 / levels);
  return {
    r: Math.min(Math.floor(r / step) * step + Math.floor(step / 2), 255),
    g: Math.min(Math.floor(g / step) * step + Math.floor(step / 2), 255),
    b: Math.min(Math.floor(b / step) * step + Math.floor(step / 2), 255),
  };
}

function toGrayscaleHex(r: number, g: number, b: number, levels: number): string {
  const gray = Math.floor(0.299 * r + 0.587 * g + 0.114 * b);
  const step = Math.floor(256 / levels);
  let val: number;

  if (levels <= 16) {
    // Match original Pic2Html pattern: 00, 11, 22, ... ff
    const nibble = Math.min(Math.floor(gray / step), 15);
    val = nibble * 17;
  } else {
    val = Math.min(Math.floor(gray / step) * step, 255);
  }

  const hex = val.toString(16).padStart(2, '0');
  return `${hex}${hex}${hex}`;
}

function toHex(n: number): string {
  return n.toString(16).padStart(2, '0');
}

// ── PNG decoder (minimal, no external deps) ──

interface ImageData {
  width: number;
  height: number;
  getPixel: (x: number, y: number) => RGB;
}

async function loadImage(path: string): Promise<ImageData> {
  // Use the canvas-less approach: decode PNG/BMP/PPM via raw parsing
  // For broad format support, shell out to ImageMagick or Python
  const ext = extname(path).toLowerCase();

  if (ext === '.ppm' || ext === '.pgm') {
    return loadPPM(path);
  }

  // For PNG/JPG/BMP/GIF — use Python+Pillow as the image decoder
  // This avoids pulling in sharp/canvas native deps
  const { execSync } = await import('child_process');
  const tmpPath = `/tmp/pic2html_pixels_${process.pid}.bin`;

  try {
    execSync(`python3 -c "
from PIL import Image
import struct, sys
img = Image.open('${path.replace(/'/g, "\\'")}').convert('RGB')
w, h = img.size
pixels = img.load()
with open('${tmpPath}', 'wb') as f:
    f.write(struct.pack('<II', w, h))
    for y in range(h):
        for x in range(w):
            r, g, b = pixels[x, y]
            f.write(bytes([r, g, b]))
"`, { stdio: 'pipe' });

    const buf = await readFile(tmpPath);
    const width = buf.readUInt32LE(0);
    const height = buf.readUInt32LE(4);
    const pixelData = buf.subarray(8);

    return {
      width,
      height,
      getPixel: (x: number, y: number): RGB => {
        const offset = (y * width + x) * 3;
        return {
          r: pixelData[offset],
          g: pixelData[offset + 1],
          b: pixelData[offset + 2],
        };
      },
    };
  } finally {
    try {
      const { unlinkSync } = await import('fs');
      unlinkSync(tmpPath);
    } catch {
      // ignore cleanup errors
    }
  }
}

async function loadPPM(path: string): Promise<ImageData> {
  const buf = await readFile(path);
  const text = buf.toString('ascii');

  // Parse PPM P6 header
  const lines = text.split('\n');
  let headerEnd = 0;
  let magic = '';
  let width = 0;
  let height = 0;
  let lineIdx = 0;

  // Skip comments, read magic, dimensions, maxval
  for (const line of lines) {
    lineIdx++;
    const trimmed = line.trim();
    if (trimmed.startsWith('#')) continue;
    if (!magic) { magic = trimmed; continue; }
    if (!width) {
      const parts = trimmed.split(/\s+/);
      width = parseInt(parts[0]);
      height = parseInt(parts[1]);
      continue;
    }
    // maxval line — skip it, then binary data follows
    headerEnd = buf.indexOf('\n', buf.indexOf(trimmed)) + 1;
    break;
  }

  const pixelData = buf.subarray(headerEnd);

  return {
    width,
    height,
    getPixel: (x: number, y: number): RGB => {
      const offset = (y * width + x) * 3;
      return {
        r: pixelData[offset],
        g: pixelData[offset + 1],
        b: pixelData[offset + 2],
      };
    },
  };
}

// ── Scaling ──

function scaleImage(img: ImageData, maxWidth: number): ImageData {
  if (img.width <= maxWidth) return img;

  const ratio = maxWidth / img.width;
  const newW = maxWidth;
  const newH = Math.floor(img.height * ratio);

  return {
    width: newW,
    height: newH,
    getPixel: (x: number, y: number): RGB => {
      const srcX = Math.floor(x / ratio);
      const srcY = Math.floor(y / ratio);
      return img.getPixel(
        Math.min(srcX, img.width - 1),
        Math.min(srcY, img.height - 1),
      );
    },
  };
}

// ── HTML generation ──

function generateHTML(img: ImageData, opts: Pic2HtmlOptions): string {
  const { size, grayscale, levels, quantize, bg } = opts;
  const lines: string[] = [];

  lines.push('<html>');
  lines.push(`<body bgcolor=${bg}>`);
  lines.push('<table border=0 cellpadding=0 cellspacing=0>');

  for (let y = 0; y < img.height; y++) {
    // Preserve the original typo at 1px — that's the fingerprint
    if (size === 1) {
      lines.push(`<tr hight=${size}>`);
    } else {
      lines.push(`<tr height=${size}>`);
    }

    for (let x = 0; x < img.width; x++) {
      const { r, g, b } = img.getPixel(x, y);
      let hex: string;

      if (grayscale) {
        hex = toGrayscaleHex(r, g, b, levels);
      } else if (quantize) {
        const q = quantizeColor(r, g, b, quantize);
        hex = `${toHex(q.r)}${toHex(q.g)}${toHex(q.b)}`;
      } else {
        hex = `${toHex(r)}${toHex(g)}${toHex(b)}`;
      }

      lines.push(`<td width=${size} bgcolor="${hex}"></td>`);
    }

    lines.push('</tr>');
  }

  lines.push('</table>');
  lines.push('</body>');
  lines.push('</html>');

  return lines.join('\n');
}

// ── Main command ──

export async function pic2html(imagePath: string, opts: Pic2HtmlOptions): Promise<void> {
  p.intro(pc.cyan('pic2html') + pc.dim(' — image to HTML table art'));

  if (!existsSync(imagePath)) {
    p.cancel(`File not found: ${imagePath}`);
    process.exit(1);
  }

  const spinner = p.spinner();
  spinner.start('Loading image...');

  let img = await loadImage(imagePath);

  if (opts.maxWidth) {
    img = scaleImage(img, opts.maxWidth);
  }

  spinner.message(`Generating ${img.width}x${img.height} HTML table...`);

  const html = generateHTML(img, opts);

  const outputPath = opts.output ||
    basename(imagePath, extname(imagePath)) + '.html';

  await writeFile(outputPath, html, 'utf-8');

  spinner.stop('Done');

  // Count unique colors
  const colors = new Set<string>();
  for (let y = 0; y < img.height; y++) {
    for (let x = 0; x < img.width; x++) {
      const { r, g, b } = img.getPixel(x, y);
      if (opts.grayscale) {
        colors.add(toGrayscaleHex(r, g, b, opts.levels));
      } else if (opts.quantize) {
        const q = quantizeColor(r, g, b, opts.quantize);
        colors.add(`${toHex(q.r)}${toHex(q.g)}${toHex(q.b)}`);
      } else {
        colors.add(`${toHex(r)}${toHex(g)}${toHex(b)}`);
      }
    }
  }

  const cells = img.width * img.height;
  const fileSize = Buffer.byteLength(html, 'utf-8');
  const mode = opts.grayscale ? 'grayscale' : opts.quantize ? 'quantized' : 'full color';

  p.note([
    `${pc.dim('Grid:')}       ${img.width} × ${img.height} (${cells.toLocaleString()} cells)`,
    `${pc.dim('Cell size:')}  ${opts.size}px`,
    `${pc.dim('Render:')}     ${img.width * opts.size} × ${img.height * opts.size}px`,
    `${pc.dim('Colors:')}     ${colors.size}`,
    `${pc.dim('Mode:')}       ${mode}`,
    `${pc.dim('Output:')}     ${outputPath} (${fileSize.toLocaleString()} bytes)`,
  ].join('\n'), 'pic2html');

  p.outro(pc.green(`✓ ${outputPath}`));
}

File: docs/research/rev-eng/pic2html/pic2html.py
#!/usr/bin/env python3
"""
pic2html — Convert images to HTML table art

Modern recreation of foxglove's 2007 Pic2Html.exe
Takes an input image and obfuscates it into an HTML table
where each pixel (or block of pixels) becomes a colored <td>.

Usage:
  python3 pic2html.py image.png                     # 1px cells, full color
  python3 pic2html.py image.png -s 24               # 24px cells (like AAAAAAAA.html)
  python3 pic2html.py image.png -s 4 -o output.html # 4px cells, named output
  python3 pic2html.py image.png --grayscale          # 16-level grayscale (like email.html)
  python3 pic2html.py image.png --grayscale -l 256   # 256-level grayscale
  python3 pic2html.py image.png --max-width 120      # scale down to 120 cells wide
  python3 pic2html.py image.png --quantize 16        # reduce to 16 colors
"""
import argparse, sys, os
from PIL import Image


def quantize_color(r, g, b, levels):
    """Snap RGB values to nearest level in a reduced palette."""
    step = 256 // levels
    r = (r // step) * step + step // 2
    g = (g // step) * step + step // 2
    b = (b // step) * step + step // 2
    return min(r, 255), min(g, 255), min(b, 255)


def to_grayscale_hex(r, g, b, levels):
    """Convert RGB to grayscale with N levels, return as hex pair repeated 3x."""
    gray = int(0.299 * r + 0.587 * g + 0.114 * b)
    step = 256 // levels
    gray = (gray // step) * step
    gray = min(gray, 255)
    # Map to the original Pic2Html pattern: 00, 11, 22, ... ff
    if levels <= 16:
        nibble = gray // 16
        val = nibble * 17  # 0→00, 1→11, 2→22, ... f→ff
    else:
        val = gray
    return f"{val:02x}{val:02x}{val:02x}"


def convert(image_path, cell_size=1, output=None, grayscale=False,
            gray_levels=16, max_width=None, quantize=None, bg="000000"):
    img = Image.open(image_path).convert("RGB")

    # Scale down if max_width specified
    if max_width and img.width > max_width:
        ratio = max_width / img.width
        new_h = int(img.height * ratio)
        img = img.resize((max_width, new_h), Image.LANCZOS)

    w, h = img.size
    pixels = img.load()

    if output is None:
        base = os.path.splitext(os.path.basename(image_path))[0]
        output = f"{base}.html"

    lines = []
    lines.append("<html>")
    lines.append(f"<body bgcolor={bg}>")
    lines.append(f"<table border=0 cellpadding=0 cellspacing=0>")

    for y in range(h):
        if cell_size == 1:
            lines.append(f"<tr hight={cell_size}>")
        else:
            lines.append(f"<tr height={cell_size}>")
        for x in range(w):
            r, g, b = pixels[x, y]

            if grayscale:
                hex_color = to_grayscale_hex(r, g, b, gray_levels)
            elif quantize:
                r, g, b = quantize_color(r, g, b, quantize)
                hex_color = f"{r:02x}{g:02x}{b:02x}"
            else:
                hex_color = f"{r:02x}{g:02x}{b:02x}"

            lines.append(f'<td width={cell_size} bgcolor="{hex_color}"></td>')

        lines.append("</tr>")

    lines.append("</table>")
    lines.append("</body>")
    lines.append("</html>")

    with open(output, "w") as f:
        f.write("\n".join(lines))

    unique_colors = set()
    for y in range(h):
        for x in range(w):
            r, g, b = pixels[x, y]
            if grayscale:
                unique_colors.add(to_grayscale_hex(r, g, b, gray_levels))
            elif quantize:
                r, g, b = quantize_color(r, g, b, quantize)
                unique_colors.add(f"{r:02x}{g:02x}{b:02x}")
            else:
                unique_colors.add(f"{r:02x}{g:02x}{b:02x}")

    file_size = os.path.getsize(output)
    print(f"pic2html — {image_path}")
    print(f"  Grid: {w}x{h} ({w * h:,} cells)")
    print(f"  Cell size: {cell_size}px")
    print(f"  Render size: {w * cell_size}x{h * cell_size}px")
    print(f"  Colors: {len(unique_colors)}")
    print(f"  Mode: {'grayscale' if grayscale else 'quantized' if quantize else 'full color'}")
    print(f"  Output: {output} ({file_size:,} bytes)")


def main():
    parser = argparse.ArgumentParser(
        description="pic2html — Convert images to HTML table art",
        epilog="Recreation of foxglove's 2007 Pic2Html.exe")
    parser.add_argument("image", help="Input image file")
    parser.add_argument("-s", "--size", type=int, default=1,
                        help="Cell size in pixels (default: 1)")
    parser.add_argument("-o", "--output", help="Output HTML file")
    parser.add_argument("--grayscale", action="store_true",
                        help="Convert to grayscale (like email.html)")
    parser.add_argument("-l", "--levels", type=int, default=16,
                        help="Grayscale levels (default: 16)")
    parser.add_argument("--max-width", type=int,
                        help="Scale image down to this many cells wide")
    parser.add_argument("--quantize", type=int,
                        help="Reduce to N color levels per channel")
    parser.add_argument("--bg", default="000000",
                        help="Page background color (default: 000000)")
    args = parser.parse_args()

    if not os.path.exists(args.image):
        print(f"Error: {args.image} not found")
        sys.exit(1)

    convert(args.image, cell_size=args.size, output=args.output,
            grayscale=args.grayscale, gray_levels=args.levels,
            max_width=args.max_width, quantize=args.quantize, bg=args.bg)


if __name__ == "__main__":
    main()
